### ISSUE TITLE
Glide docking with Maize gives "Unknown job querying" error

### ISSUE THREAD

**Issue by maizerunners** [MolecularAI/maize#12]
Hi!

I've been trying to implement Glide docking as a scoring component in reinvent4 using Maize (staged learning). The first step of the reinvent workflow works as intended, generating molecules based on 2D scoring components. In the second stage, Maize is launched and starts Schrodinger Ligprep with the input.csv (SMILES) provided by reinvent. I can see that Maize launches Ligprep, and the first batch of molecules are prepared in the "node-ligprep" node, but the workflow gets stuck with the error message "Unknown job querying error". I'm running locally on a workstation without slurm, but none of the host settings I've tried so far have worked. It does not appear to be a problem with Schrodinger, since both the Ligprep and Glide modules seem to work as intended with jobcontrol.

If this is a known issue and anyone knowns how to respolve it, I would be very grateful.

Configuration files for reinvent and Maize (+ the maize log) are attached for reference.

[maize_worfklow_docking.log](https://github.com/user-attachments/files/21936799/maize_worfklow_docking.log)

[glide.yml](https://github.com/user-attachments/files/21936805/glide.yml)

--------------------------------------MAIZE SCORING COMPONENT FOR REINVENT-------------------------------------------------

[[component]]
[component.Maize]

[[component.Maize.endpoint]]
name = "docking with MAIZE"
weight = 1
params.executable = "/home/reinvent/.conda/envs/reinvent4/bin/maize"
params.workflow = "/home/reinvent/PROJECTS/Myproject/glide.yml"
params.config = "/home/reinvent/.config/maize.toml"
params.log = "/home/reinvent/tmp/maize_worfklow_docking.log"
transform.type = "reverse_sigmoid"
transform.high = -6
transform.low = -13.5
transform.k = 0.2
inp_grid      = "/home/reinvent/PROJECTS/Myproject/glide-grid_test/glide-grid_test.zip"
save_location = "/home/reinvent/PROJECTS/Myproject/glide_runs/stage2"
save_poses    = true
n_jobs        = 1
host          = "Olsson-alpha"
-----------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------REINVENT.toml----------------------------------------------------------------------------------------

### REINVENT4 TOML input example for reinforcement/curriculum learning

run_type = "staged_learning"
device = "cuda:0"
tb_logdir = "/home/reinvent/PROJECTS/Myproject/tb_logs"  

[parameters]
summary_csv_prefix = "/home/reinvent/PROJECTS/Myproject/reinvent_results"  # prefix for the CSV file

## Reinvent
prior_file = "/home/reinvent/PROJECTS/Myproject/reinvent.prior"
agent_file = "/home/reinvent/PROJECTS/Myproject/reinvent.prior"
batch_size = 4          # network

[learning_strategy]
type = "dap"      # dap: only one supported
sigma = 128       # sigma of the RL reward function
rate = 0.0001     # for torch.optim

[diversity_filter]  # optional, comment section out or remove if unneeded
                    # NOTE: also memorizes all seen SMILES
type = "IdenticalMurckoScaffold" # IdenticalTopologicalScaffold,
                                 # ScaffoldSimilarity, PenalizeSameSmiles
bucket_size = 25                 # memory size in number of compounds
minscore = 0.4      # only memorize if this threshold is exceeded

### Stage1
[[stage]]

chkpt_file = '/home/reinvent/PROJECTS/Myproject/stage_1.chkpt'  # name of the checkpoint file, can be reused as agent
termination = "simple"  # termination criterion fot this stage
max_score = 0.2  # terminate if this total score is exceeded
min_steps = 20  # run for at least this number of steps
max_steps = 30  # terminate entire run when exceeded

[stage.scoring]  # the scoring components can be read from a score file
type = "geometric_mean"  # aggregation function
filename = "/home/reinvent/PROJECTS/Myproject/stage1_scoring.toml"  # file with scoring setup for this stage
filetype = "toml"  # file format: TOML or JSON, no default, must be present

### Stage2
[[stage]]

chkpt_file = '/home/reinvent/PROJECTS/Myproject/stage_2.chkpt'  # name of the checkpoint file, can be reused as agent
termination = "simple"  # termination criterion fot this stage
max_score = 0.3  # terminate if this total score is exceeded
min_steps = 5  # run for at least this number of steps
max_steps = 10  # terminate entire run when exceeded

[stage.scoring]  # the scoring components can be read from a score file
type = "geometric_mean"  # aggregation function
filename = "/home/reinvent/PROJECTS/Myproject/stage2_scoring.toml"  # file with scoring setup for this stage
filetype = "toml"  # file format: TOML or JSON, no default, must be present

---------------------------------------------------------MAIZE.toml (config)----------------------------------------------------------------------

# maize global configuration file example

# Where to save temporary files and all workflow directories
scratch = "/home/reinvent/tmp"

# Additional packages to load
packages = [
    "maize.steps.mai",
    "maize_contrib.reinvent",
    "maize_contrib.glide",
    "maize_contrib.filters",
    "maize.graphs.mai"
]

# Environment variables to be set globally
[environment]
SCHRODINGER = "/opt/schrodinger2025-1"
SCHROD_LICENSE_FILE = "/opt/schrodinger/licenses/schrod1_27045.serverid.lic"

#Tool-specific_section
[ligprep]
commands.ligprep = "/opt/schrodinger2025-1/ligprep"
type = "Ligprep"
host = "Olsson-alpha"
keywords = { ATTACHED = true }

[glide]
commands.glide = "/opt/schrodinger2025-1/glide"
type = "Glide"
host = "Olsson-alpha"
precision = "SP"

[batch]
system = "local"  # Can be one of {'cobalt', 'flux', 'local', 'lsf', 'pbspro', 'rp', 'slurm'}
max_jobs = 1  # The maximum number of jobs that can be submitted by a node at once

**Comment by marcobICR** [comment]
I'm not from AZ, but I think that this issue would be best posted in the [maize-contrib ](https://github.com/MolecularAI/maize-contrib/tree/63c02de4393374a98b2d25b9944a26c2a64328c3)repository.

Anyway, Maize's Glide node assumes that you are using the newer Schrodinger JobServer, not JobControl. It will check for a JobServer, and if there's none it will create a Local JobServer to handle the jobs. 
However, if you are still using the older JobControl manager for other tasks, you probably still have the JobServer usage flag deactivated.

You can check with this command:
```$SCHRODINGER/utilities/feature_flags -l``` 
and you should see something like this if it's disabled:

```JOB_SERVER : **Disabled**
    User State    : Enabled
    Default State : Disabled
    Source        : $USER/.schrodinger/featureflags.json```

If it's disabled, try using this command to activate it **before running maize**:

```$SCHRODINGER/utilities/feature_flags -e JOB_SERVER``` 

If you want to use schrodinger in some other way, remember to deactivate it by using

```$SCHRODINGER/utilities/feature_flags -d JOB_SERVER``` 


Let me know if you managed to solve the issue!

**Comment by maizerunners** [comment]
Hi @marcobICR,

Thank you for the help! The activation didn't quite fix it, but it was related to the JobServer as you wrote. However, I got some help by people at AZ to instead disable the JobServer feature in maize-contrib/maize/steps/mai/common/schrodinger.py, and now it works well!

**Comment by marcobICR** [comment]
That's great! Could you share the fix you used please?

**Comment by maizerunners** [comment]
Sure!

In maize-contrib/maize/steps/mai/common/schrodinger.py, we deleted everything between lines 408-444. We then added "self.logger.debug(f"Job result {result}") under "result = cmd.run_only(command, verbose=verbose)" on line 407.

# Before
       `# Submit job and check for errors upon submission
        self.logger.debug("Running Schrodinger job '%s'", " ".join(command))
        cmd = CommandRunner(raise_on_failure=False, working_dir=working_dir, validators=validators)
        result = cmd.run_only(command, verbose=verbose)

        match _query_schrodinger_submission(result):
            case _SchrodingerSubmissionStatus.SERVER:
                self.logger.warning(
                    "Job server communication failure, restarting jobserver\n %s",
                    result.stdout.decode(),
                )
                self._restart_jobserver()
                time.sleep(5)
                return _restart(n_fail=_n_failures + 1)

            case _SchrodingerSubmissionStatus.HOST:
                self.logger.warning(
                    "Host %s is unavailable, you may need to run 'jsc cert get <host>:<port>'",
                    self.host.value,
                )
                if self.fallback.value and self.host.value != "localhost":
                    n_jobs = min(self.n_jobs.value, cpu_count())
                    self.host.set("localhost")
                    self.logger.warning(
                        "Falling back to 'localhost', using %s jobs (instead of %s jobs)",
                        n_jobs,
                        self.n_jobs.value,
                    )
                    self.n_jobs.set(n_jobs)
                    return _restart(n_fail=0)
                check_returncode(result, raise_on_failure=raise_on_failure, logger=self.logger)
                self._cleanup_jobserver_temp()
                return result

            case _SchrodingerSubmissionStatus.OTHER:
                self.logger.debug("Other failure, returncode %s", result.returncode)
                check_returncode(result, raise_on_failure=raise_on_failure, logger=self.logger)
                self._cleanup_jobserver_temp()
                return result

            case _SchrodingerSubmissionStatus.SUCCESS:
                self.logger.debug("Submission succeeded, now monitoring...")

       # Submission success, get the ID and monitor progression`

# After
        `# Submit job and check for errors upon submission
        self.logger.debug("Running Schrodinger job '%s'", " ".join(command))
        cmd = CommandRunner(raise_on_failure=False, working_dir=working_dir, validators=validators)
        result = cmd.run_only(command, verbose=verbose)
        self.logger.debug(f"Job result {result}")

        # Submission success, get the ID and monitor progression`
